service/worker/scanner/history/scavenger.go

```diff
@@ -105,186 +107,223 @@ func NewScavenger(
 	logger log.Logger,
 ) *Scavenger {
 
-	rateLimiter := rate.NewLimiter(rate.Limit(rps), rps)
-
 	return &Scavenger{
-		db:      db,
-		client:  client,
-		hbd:     hbd,
-		rps:     rps,
-		limiter: rateLimiter,
+		db:     db,
+		client: client,
+		rateLimiter: quotas.NewDefaultOutgoingDynamicRateLimiter(
+			func() float64 { return float64(rps) },
+		),
 		metrics: metricsClient,
 		logger:  logger,
+
+		hbd: hbd,
 	}
 }
 
 // Run runs the scavenger
 func (s *Scavenger) Run(ctx context.Context) (ScavengerHeartbeatDetails, error) {
-	taskCh := make(chan taskDetail, pageSize)
-	respCh := make(chan error, pageSize)
-	concurrency := s.rps/rpsPerConcurrency + 1
+	reqCh := make(chan taskDetail, pageSize)
 
-	for i := 0; i < concurrency; i++ {
-		go s.startTaskProcessor(ctx, taskCh, respCh)
+	for i := 0; i < numWorker; i++ {
+		s.WaitGroup.Add(1)
+		go s.taskWorker(ctx, reqCh)
 	}
+	s.WaitGroup.Add(1)
```

Comment:
what value is this additional Add(1) providing? The task workers anyway won't exit until loadTasks finishes, right?

Reply:
down below the logic do wait?
```go
s.WaitGroup.Wait()
```